Fix 128bit SimHash, and add 256bit and 512bit SimHash

nicolaichuk opened the pull request:
@tgalopin, hi.
Please review and merge to upstream.

nicolaichuk committed:
fix 128 bit SimHash
add 256 and 512 bit SimHash

nicolaichuk committed:
small bug fix

tgalopin commented on lib/Tga/SimHash/Comparator/GaussianComparator.php:
```diff
@@ -36,6 +36,28 @@ public function __construct($deviation = 4)
+     * @param Fingerprint $fp2
+     * @return int
+     */
+    private static function countDifferences(int $size, Fingerprint $fp1, Fingerprint $fp2) {
```
I'm pretty sure this method is slower than the current one, why do you need this?

nicolaichuk replied:
Because in the current implementation of 128bit version, there is a loss of precision in the conversion.

tgalopin replied:
Got it, thanks :) .

tgalopin replied:
You can't use scalar type hint, we support php 5.3+

nicolaichuk replied:
remove  scalar type hint

tgalopin commented on lib/Tga/SimHash/Comparator/GaussianComparator.php:
```diff
@@ -36,6 +36,28 @@ public function __construct($deviation = 4)
     }
 
     /**
+     * Count Differences beetwin fingerprints.
```
`Count differences between fingerprints`

nicolaichuk replied:
fix

tgalopin commented on lib/Tga/SimHash/Fingerprint.php:
```diff
@@ -24,19 +24,19 @@ class Fingerprint
      * @var float
      */
-    protected $decimalValue;
+    protected $value;
```
This is a BC break without any added value, could you revert this?

nicolaichuk replied:
"decimalValue" save value as number.
"value" save  value as string.

These are two different variables, so they are called differently.

tgalopin replied:
Could you make this property private as well? I don't think it should be exposed anyway.

tgalopin commented on lib/Tga/SimHash/Tokenizer/String128Tokenizer.php:
```diff
@@ -18,13 +18,20 @@
     public function tokenize($element)
     {
-        return str_pad(base_convert(md5($element), 16, 2), 128, '0', STR_PAD_LEFT);
+        $hash = md5($element);
```
Why do you need to split this in parts?

nicolaichuk replied:
For more comfort debugging

tgalopin commented on lib/Tga/SimHash/Tokenizer/String128Tokenizer.php:
```diff
@@ -18,13 +18,20 @@
 class String128Tokenizer implements TokenizerInterface
 {
+ 
+    protected static $search = array('0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f');
```
You shouldn't add protected variables like this without a specific need as they will need to be handled by BC promise.

nicolaichuk replied:
fix

tgalopin commented on lib/Tga/SimHash/Tokenizer/String256Tokenizer.php:
```diff
@@ -0,0 +1,58 @@
+    {
+        return $size === 256;
+    }
+}
```
Please add an ending line to this file.

nicolaichuk replied:
fix

tgalopin commented on lib/Tga/SimHash/Tokenizer/String512Tokenizer.php:
```diff
@@ -0,0 +1,58 @@
+    {
+        return $size === 512;
+    }
+}
```
Please add an ending line to this file.

nicolaichuk replied:
fix

tgalopin reviewed (changes requested):
Hello @nicolaichuk,

Thanks for your PR!

I'm not sure what you want to fix in 128 bits version and why you need all the modifications you did. Unfortunately, I can't accept such a PR without some explanations about your work and fixes on the coding style and BC breaks.

nicolaichuk committed:
* change comment
* add an ending line to this file

nicolaichuk commented:
@tgalopin

> I'm not sure what you want to fix in 128 bits version and why you need all the modifications you did. 

first:
"base_convert" in 128 bits version always return string content only zero symbols.
(environment: php 7.1, windows 7).

second:
php "bindec" dont corect work with 128-bit value 
how you try safe 128-bit value in 32-bit float?

nicolaichuk committed:
change protected to private

tgalopin commented on lib/Tga/SimHash/Tokenizer/String128Tokenizer.php:
```diff
@@ -18,13 +18,20 @@
  */
 class String128Tokenizer implements TokenizerInterface
 {
+ 
```
Please remove this extra line.

nicolaichuk replied:
fix

tgalopin commented on lib/Tga/SimHash/Tokenizer/String256Tokenizer.php:
```diff
@@ -0,0 +1,58 @@
+ */
+class String256Tokenizer implements TokenizerInterface
+{
+ 
```
Please remove this extra line.

nicolaichuk replied:
fix

tgalopin commented on lib/Tga/SimHash/Tokenizer/String512Tokenizer.php:
```diff
@@ -0,0 +1,58 @@
+ */
+class String512Tokenizer implements TokenizerInterface
+{
+ 
```
Please remove this extra line.

nicolaichuk replied:
fix

tgalopin reviewed (commented):
Thanks for your work, I understand it now :) . A few comments and we should be okay :) !

nicolaichuk committed:
fix by code review https://github.com/tgalopin/SimHashPhp/pull/5

nicolaichuk commented:
@tgalopin, ping...

tgalopin reviewed (changes requested):
Travis is failing due to scalar type hints.

bbalet commented on lib/Tga/SimHash/Tokenizer/String128Tokenizer.php:
```diff
@@ -18,13 +18,19 @@
-        return str_pad(base_convert(md5($element), 16, 2), 128, '0', STR_PAD_LEFT);
+        $hash = md5($element);
+        $hash = str_replace(self::$search, self::$replace, $hash);
+        $hash = str_pad($hash, 512, '0', STR_PAD_LEFT);
```
I think there is a typo here. Shouldn't you write:

    $hash = str_pad($hash, 128, '0', STR_PAD_LEFT);

nicolaichuk replied:
@bbalet, fixed.

bbalet commented:
Hi guys,

I know what is to maintain an OSS project, but this PR looks good to me except a typo and the scalar type hint (except if you want to stop to maintain compatibility with older projects).

It is a good lib that helped me in the past, but now I am stuck with the error in 128 hashes.

How can I help?

tgalopin commented:
Hello @bbalet !

The issue is that as @nicolaichuk didn't gave me access to code modification in this PR, I can't update his code, and using it without crediting him does not feel right in open-source :) . If you want to open a new PR based on his code, fixing the issues and adding a note in your commit message about @nicolaichuk, please don't hesitate to do so and I will review it happily :) !

nicolaichuk commented:
@tgalopin
>> he issue is that as @nicolaichuk didn't gave me access to code modification in this PR

in this PR check "Allow edits from maintainers. ".

nicolaichuk committed:
support php 5.3+

nicolaichuk committed:
fix function call
by code review @bbalet https://github.com/tgalopin/SimHashPhp/pull/5#discussion_r211200121
